src/lib/warn-upload-overwrite.js

```diff
@@ -0,0 +1,184 @@
+  } catch (e) {
+    // continue regardless of error
+    log.trace('Trying to fetch local _rev', e);
+  }
+
+  if (localRev) {
+    localDoc._rev = localRev;
+  }
+
+  if (localDoc._attachments) {
+    for (let key in localDoc._attachments) {
+      localDoc._attachments[key] = {
+        'content_type': doc._attachments[key].content_type, 
+        'revpos': localRev ? parseInt(localRev.split('-')[0]) : 0, 
+        'length': doc._attachments[key].data.length,
+        'digest': `md5-${crypto.createHash('md5').update(doc._attachments[key].data, 'binary').digest('base64')}`,
+        'stub': true
+      };
+    }
+  }
+
+  let diff; 
+  if (localDoc._id === 'settings' && remoteDoc._id === 'settings') {
+    diff = jsonDiff.diffString(remoteDoc.settings, localDoc.settings);
+  } else {
+    diff = jsonDiff.diffString(remoteDoc, localDoc);
+  }
+  if (diff) {  
+    if (!localRev) {
+      if (!readline.keyInYN(`We can't determine if you're going to overwrite someone else's changes or not. Would you like to proceed?`)) {
+        throw new Error('configuration modified');
+      }
+    } else if (localRev !== remoteRev) {
+      let index = readline.keyInSelect(responseChoicesWithDiff, question, {cancel: false});
+      if (index === 2) { // diff
+        log.info(diff);
+
+        index = readline.keyInSelect(responseChoicesWithoutDiff, question, {cancel: false});
+      }
+      if (index === 1) { // abort
+        throw new Error('configuration modified');
+      }
+    }
+  }
+
+  return doc;
+};
+
+const postUploadByRev = async (db, doc) => {
+  const remoteDoc = await db.get(doc._id);
+
+  const revsDir = `${environment.pathToProject}/._revs`;
+  if (!fs.exists(revsDir)){
+    fs.mkdir(revsDir);
+  }
+
+  let revs = {};
+
+  const revsFile = `${revsDir}/${doc._id}.json`;
+  if (fs.exists(revsFile)) {
+    Object.assign(revs, JSON.parse(fs.read(revsFile).trim()));
+  }
+  revs[getRevsAndHashesDocKey()] = remoteDoc._rev;
+
+  fs.write(revsFile, JSON.stringify(revs));
+
+  return doc;
+};
+
+const preUploadByXml = async (db, docId, localXml) => {
+  let storedHash;
+  try {
+    storedHash = JSON.parse(fs.read(`${environment.pathToProject}/.hashes/${docId}.json`).trim())[getRevsAndHashesDocKey()];
+  } catch (e) {
+    // continue regardless of error
+    log.trace('Trying to fetch stored hash', e);
+  }
+
+  const localHash = crypto.createHash('md5').update(localXml, 'binary').digest('base64');
+  if (storedHash === localHash) {
+    throw new Error('No changes');
+  }
+
+  let remoteXml;
+  try {
+    const buffer = await db.getAttachment(docId, 'xml');
+    remoteXml = buffer.toString('utf8');
+  } catch (e) {
+    // continue regardless of error
+    log.trace('Trying to fetch remote xml', e);
+    throw e;
+  }
+
+  const remoteHash = crypto.createHash('md5').update(remoteXml, 'binary').digest('base64');
+  if (storedHash === remoteHash) {
+    throw new Error('No changes');
```

Comment:
Doesn't this mean that the modified local xml won't be uploaded? In this case, we know the local xml has been modified and should be uploaded. We also know that the configuration was up to date before uploading. So there's no conflict, there is a change, we don't want to prompt the user we just want to upload the local xml and update the stored hash. This is the best case scenario!

I think we should `return Promise.resolve()` here, right?

Reply:
Oh yes, you are right! 👍